### ISSUE TITLE
[google_sign_in_web] Display Sign In With Google button

### ISSUE THREAD

**Issue by c-seeger** [flutter/flutter#121781]
## Use case

Using `google_sign_in_web` 6.0.0+ currently needs `signInSilently` to be triggered on page initialization. 

If your login page contains multiple login options, a request is made by `singInSilently` before the user decides what login option to use.

The current sugguestion in this case is:

[Display the Sign In With Google button](https://developers.google.com/identity/gsi/web/guides/display-button)

Which is currently not implemented on `google_sign_in_web`

Discovered as part of: #88084

## Proposal

Add the "Sign In With Google Button" option:

- trigger google sign in services on button press and not in advance
- beeing able to customise the button to match style with other login options

![](https://developers.google.com/static/identity/gsi/web/images/personalized-button-single.png)

**Comment by c-seeger** [comment]
@ditman would this work out of the box once the internal issue is solved you mentioned [here](https://github.com/flutter/flutter/issues/88084#issuecomment-1439088109)?

If `signIn` returns the idToken (as before 6.0.0) it would be the same usage as before or did I missed sth?

**Comment by ditman** [comment]
> If signIn returns the idToken (as before 6.0.0) it would be the same usage as before or did I missed sth?

Yes, if the plugin's `signIn` was always able to return the `idToken`, this button wouldn't be "needed".

This button is needed in the current SDK to have a reliable way of getting an idToken via a popup, rather than just relying on the OneTap UX, or to render this on pages where the OneTap UX doesn't make much sense (like a login page).

I have already implemented the [JS-interop Dart API, here](https://github.com/flutter/packages/commit/b6fe67e49fc280bfdde1a7e36b5b39e8661f06de).

In order to fix this, we need to:

* Add a new Widget to the `google_sign_in_web` package that:
  * Uses a platform view to configure a DIV where the SDK can render the button (possibly with fixed size), then:
  * Uses the API I linked above to instruct the JS SDK to render the button inside the platform view.

I'll try to prototype this tomorrow.

(We also probably need to expose to the public a `hasAuthorizedScopes([scopes])` method to the core plugin, to start separating authentication checks from authorization checks.)

**Comment by ditman** [comment]
I did some digging here, and because of the way the core plugin connects to the implementations there's no easy way to convert this into a "web-only" change; since there's no way to inject "new user credentials" from the web version to the plugin core; this is the only place from where we can add user credentials to the current user: 

https://github.com/flutter/packages/blob/405e4652708b82470528b9eebee6e758bb4caadb/packages/google_sign_in/google_sign_in/lib/google_sign_in.dart#L284

Even if this is a web-only Widget, we must expose it through the Core plugin so we can cause an effect in the last-seen user.

We were going to change the platform_interface anyways to add a few extra methods, so this is only delaying the PoC a little bit.

My current operating assumptions are that I want to expose the following from the implementation plugins:

* A `Stream<GoogleSignInAccount?>` so the implementation can directly notify the core plugin if authentication has changed. This will allow us to implement not only the async authentication from the custom button, but also we can timeout the current user when their idToken expires without the plugin needing to contain that logic. This stream would be directly re-broadcast from the `onCurrentUserChanged` stream.
* A `Future<bool> hasGrantedScopes(List<String> scopes)` or `Future<bool> canAccessScopes(String? accessToken, List<String> scopes)` so users can validate that their `accessToken` can still access the `scopes` passed in. This would need to validate both the expiration of the token, and also that the user has previously granted enough permissions.

Once the two methods above are implemented and demoed, we can change the `silentSignIn` method on the web to start returning only the idToken (if any), now that programmers have the option to check if the Authorization of a user is complete or not.

The flow would be something like this, I think (marking new components with `*`):

```mermaid
flowchart TD
    a[Render Page] --> A
    A(signInSilently) --> B(onCurrentUserChanged - may be null)
    c[*Web Button Widget] --> d
    d[*Web user stream] --> B
    B --> C{isSignedIn}
    C -->|Yes| D{*canAccessScopes}
    D -->|Yes| Z[Done!]
    D -->|No| E(requestScopes)
    E --> D
    C -->|No| F(signIn)
    F --> B
```

For **existing apps that need to support the web**, the only difference is that "isUserAuthenticated" and "canAccessScopes" now need to be checked explicitly, and separately.

On mobile, we can just use the same assumptions that were being made before. `canAccessScopes` is the same as `isSignedIn`, and their "user stream" would always be an empty stream.

_A change that would benefit all apps would be to **only rely on the `onCurrentUserChanged` stream**, and treat `signIn` and `silentSignIn` as if they always returned `Future<void>`. We can modify the `example` app to illustrate this pattern, but it'd be **very breaking** to change this across the whole ecosystem._

**Comment by Levi-Lesches** [comment]
My two cents: given that Google is outright discontinuing the APIs this month, I think it's okay for the Flutter side to push for breaking changes that streamline the transition, rather than a confusing two-step migration that isn't _technically_ breaking but forces devs to think about and use the APIs in new ways anyhow.

**Comment by ditman** [comment]
>  I think it's okay for the Flutter side to push for breaking changes that streamline the transition, rather than a confusing two-step migration that isn't technically breaking but forces devs to think about and use the APIs in new ways anyhow.

@Levi-Lesches that's good feedback; I went back and forth with some of the new API behaviors being different in the latest version of the web plugin (do I make it fully breaking or not that much).

The problem that I have is that I don't have the bandwidth to do a full-blown plugin from scratch (because for mobile nothing changes), so I need to update the one we have, as we need new features or we get them from the JS SDK.

**Comment by ditman** [comment]
I have a small PoC here:

* https://dit-gis-test.web.app

I have implemented more or less what I described above to enable the "Button" widget in the web. I like the "rx" way of using the `onCurrentUserChanged` Stream as an only source of truth for signed-in / out states on apps.

The changes to the example app are not very large. The biggest of which is adding an extra `canAccessScopes` check that adds an extra click to the authentication process. (Now that `canAccessScopes` exists, we can return an user from the signInSilently method too!)

Next steps for this task is to clean-up/reduce the code as much as possible, and add tests. Then add features to the Button Widget, like the ability to customize its appearance/size, and attempt to make it animate and reflow the website around it (this one might be trickier!)

**Comment by c-seeger** [comment]
Great job so far. 

Tested it with chrome on linux also with inkognito mode. Both worked as expected. On first time page load I get the google popup asking if i want to login. If i cancel the popup and just use the button everything is fine and on followup pageloads the popup does not appear anymore. On Inkognito mode no popup appears and button solution works also.

If I understand the flow correctly it should be possible to only implement the "left" or "right" part of the flow so either you use `signInSilently` to trigger the flow below `onCurrentUserChanged` or the web button. I guess your example app implements both currently? (Thats why the popup shows up on first page load?). :+1: 

I'm also with @Levi-Lesches, I don't see an issue with having a breaking change if it makes the library more compact and easier to maintain in the long run it might be better than building workarounds to keep code compatibility. 
Since google announced this change anyway at least from my side it was expected that we need to change the code to some degree.

**Comment by SeriousMonk** [comment]
> I have a small PoC here:
> 
> * https://dit-gis-test.web.app
> 
> I have implemented more or less what I described above to enable the "Button" widget in the web. I like the "rx" way of using the `onCurrentUserChanged` Stream as an only source of truth for signed-in / out states on apps.
> 
> The changes to the example app are not very large. The biggest of which is adding an extra `canAccessScopes` check that adds an extra click to the authentication process. (Now that `canAccessScopes` exists, we can return an user from the signInSilently method too!)
> 
> Next steps for this task is to clean-up/reduce the code as much as possible, and add tests. Then add features to the Button Widget, like the ability to customize its appearance/size, and attempt to make it animate and reflow the website around it (this one might be trickier!)

Awesome! I really need this to keep the authentication on my website working after the migration. The only thing I need is the idToken, to send it to my backend for validation, data extraction and user authentication. For now there seems no way of achieving this so this widget would really come in clutch!

Will this be ready before the 31st Match?

**Comment by ditman** [comment]
> If I understand the flow correctly it should be possible to only implement the "left" or "right" part of the flow so either you use signInSilently to trigger the flow below onCurrentUserChanged or the web button. I guess your example app implements both currently? (Thats why the popup shows up on first page load?). 👍

Yes, I think the idea of the `OneTap` experience is to have it trigger *everywhere* on your app so it's very easy for users to authenticate (it might not show up if the user manually closes it), and then have a "login" page that renders the button, but not the `OneTap`.

The problem is that our demo app starts directly in a Login page :)

There's also an internal bug requesting (begging) for the idToken in the current Oauth flow, and that would give us the same behavior as before (idToken + accessToken in a single step), but until the JS SDK gives us that, we cannot guarantee that the "signIn" method in the web will return an idToken :/

**Comment by ditman** [comment]
I've updated the PoC with configuration options for the widget (those will be removed from the "example" app before I publish this, for sure)

* https://dit-gis-test.web.app

I've published my progress so far to [a branch](https://github.com/flutter/packages/compare/main...ditman:flutter-packages:gsi-web-add-button). It needs a bunch of tests (and to go through the review process ;) )

**Comment by c-seeger** [comment]
Took a short review, looks great so far :+1: . 

I've seen we are able to change a lot of things like ButtonTheme, etc. This looks good. 

Maybe a few things that might be handy:

Is it possible to interchange the logo or disable it? 

Also I've noticed: 

```
enum ButtonText {
  /// The button text is "Sign in with Google".
  signinWith,
  /// The button text is "Sign up with Google".
  signupWith,
  /// The button text is "Continue with Google".
  continueWith,
  /// The button text is "Sign in".
  signin;
}
```
is set to fixed values.

Is it possible to set the ButtonText freely? This might can cause issues when using different languages. For now it looks like the translation is done by using the browser language. This might conflict if the app just uses it's own settings to set the language default. As an alternative an option to set the translation language manually might also work.

Example of our current logo tower:

![logo-tower](https://user-images.githubusercontent.com/68909526/224475028-dc4fb48b-e183-45ad-8b97-93987cdfd28e.png)

It doesn't have to look the same after refactoring we can change the style of the other buttons to fit the color filled rectengular logo version, at least it will work as long as facebook or apple don't come up with the same idea to design their own buttons :P, then things can get a bit more tricky to match styles. But so far this is not the case.

**Comment by ditman** [comment]
@c-seeger unfortunately, the customization options that we have are the same as those of the JS SDK, so responding to your questions:

> Is it possible to interchange the logo or disable it?

Nope, it seems the logo will always show up. You can specify it to be `logoAlignment: centered`, rather than `left`, if you specify a `minimumWidth` that is wider than the actual button (max 400px wide)

> Is it possible to set the ButtonText freely? 

No, there's only the 4 options presented there.

> This might can cause issues when using different languages. This might conflict if the app just uses it's own settings to set the language default. As an alternative an option to set the translation language manually might also work.

There's a `locale` option that can be set when rendering the button for it to match your site. [Docs](https://developers.google.com/identity/gsi/web/reference/js-reference#locale). I cannot find the list of supported locales, though.

Here are all the options:

* https://developers.google.com/identity/gsi/web/reference/js-reference#GsiButtonConfiguration

**Comment by c-seeger** [comment]
Thanks @ditman, for taking the time. The `locale` option is nice. The rest is more eye candy we can work around. Will give it a try and see if i can get it working.

**Comment by c-seeger** [comment]
Gave it a try but run into one issue I'm not sure how to solve.

Dropped the example in this [gist](https://gist.github.com/c-seeger/0dfe2268b734600fb9ed08c3756db3ed)

The easy one is `titleTextStyle` within `ListTile` in `web_eye_candy.dart` is unknown (I'm using the latest flutter version but maybe you had some dev version where this was added, just outcommented it since it's just eye candy for now)

The tricky issue I wasnt able to solve is 

```
../../../../../../../.pub-cache/git/flutter-packages-aa14d333c7ca985d5b4142d7f9a4ecec37762b10/packages/google_sign_in/google_sign_in_web/lib/src/flexible_size_html_element_view.dart:32:50: Error: The method 'allowInterop' isn't defined for the class
'_FlexHtmlElementView'.
 - '_FlexHtmlElementView' is from 'package:google_sign_in_web/src/flexible_size_html_element_view.dart'
```

Looks like for some reason the JS can't be mapped correctly using allowInterop. Not sure what I'm missing here.

**Comment by ditman** [comment]
@c-seeger `allowInterop` should come from the `import 'package:js/js_util.dart';`. In Dart prior to 3, it may come from `package:js/js.dart` only. (I'm using a fairly recent version of Flutter `master` to develop this, and still haven't tried it with older versions!)

**Comment by ditman** [comment]
@c-seeger I've restructured the code so (I hope) it's easier to follow and test. The `google_sign_in` example app now focuses only on the changes to allow the web behavior alongside the mobile one, and the button customization bits are a separate example app in the `google_sign_in_web` package.

I've deployed the two example apps:

* https://dit-gis-test.web.app (Google Sign In example)
* https://dit-gis-button-test.web.app (Sign in with Google button Tester)

I've also created a draft PR so it's easier to comment on the code, rather than having to reference code from this issue to a branch.

* https://github.com/flutter/packages/pull/3461

_(In the process I completely destroyed and re-created my branch, so my apologies if you forked it and worked on top of my changes, you might experience some breakage!)_

**Comment by wreppun** [comment]
I've been trying to wade through the changes and the docs (from developers.google.com), so apologies if I've missed this.

Am I correct that "Button" sign in has to land before we can get an `idToken` for a fresh user (fresh as in not-already-signed-into-the-browser)?

`signInSilently` calls `id.prompt(...)` under the hood, and if there is no existing session, we get a `opt_out_or_no_session`.  _OneTap_ is only designed to work with users who are already logged into a google account in the browser.

**Comment by ditman** [comment]
> Am I correct that "Button" sign in has to land before we can get an idToken for a fresh user (fresh as in not-already-signed-into-the-browser)?

@wreppun This is correct.

 > signInSilently calls id.prompt(...) under the hood, and if there is no existing session, we get a opt_out_or_no_session. OneTap is only designed to work with users who are already logged into a google account in the browser.

Yes. Their browser session doesn't have to come from your site, however, if they sign in into gmail (for example), OneTap should work too, but I understand how this is Not Desirable.

**Comment by meansoup** [comment]
@ditman Do you have plan to deploy new version? include this fixes?

**Comment by ditman** [comment]
@meansoup yes, my changes have passed the first round of code-review, so I'll split it into the 3 different code reviews needed on Monday.

**Comment by ditman** [comment]
All right, the final piece of the puzzle is about to land!

* https://github.com/flutter/packages/pull/3642

In `google_sign_in: ^6.1.0`:

* `signIn` on the web is discouraged, and will `warn` you in development if you attempt to use it.
* `signInSilently` on the web returns non-null users again! (Those only have an `idToken`).
* `Widget renderButton(options)` may result in an `onCurrentUserChanged` event when a user successfully authenticates (`idToken` only).
  * This method is web-only, so if your app needs to support mobile as well, you'll have to call it [using Dart "conditional exports"](https://github.com/ditman/flutter-packages/blob/54114b22d981604a4a382a6c4ff643aa5889e4bf/packages/google_sign_in/google_sign_in/example/lib/src/sign_in_button.dart), then [used like so](https://github.com/ditman/flutter-packages/blob/54114b22d981604a4a382a6c4ff643aa5889e4bf/packages/google_sign_in/google_sign_in/example/lib/main.dart#L209-L213).
* In the web, your app [must check `canAccessScopes(scopes)`](https://github.com/ditman/flutter-packages/blob/54114b22d981604a4a382a6c4ff643aa5889e4bf/packages/google_sign_in/google_sign_in/example/lib/main.dart#L58-L63) for the scopes it needs, other than `email`, `profile` and `openid`, (which seem to come for free with the Sign In token).
  * And have a way to [request more scopes](https://github.com/ditman/flutter-packages/blob/54114b22d981604a4a382a6c4ff643aa5889e4bf/packages/google_sign_in/google_sign_in/example/lib/main.dart#L147-L161) from a [button tap](https://github.com/ditman/flutter-packages/blob/54114b22d981604a4a382a6c4ff643aa5889e4bf/packages/google_sign_in/google_sign_in/example/lib/main.dart#L147-L161).
    * This should produce a new `accessToken` when used.

Anyway, please do take a look at what's coming. 

I've created an issue so the community can gather and discuss different integration techniques for the above (maybe for people who use Bloc or other state management systems) after this issue closes:

* https://github.com/flutter/flutter/issues/124297

**Comment by ditman** [comment]
google_sign_in 6.1.0 is now [published](https://pub.dev/packages/google_sign_in/changelog#610)!
